frontend/src/metabase/dashboard/actions/getNewCardUrl.ts

```diff
@@ -0,0 +1,110 @@
+import _ from "underscore";
+
+import { getCardAfterVisualizationClick } from "metabase/visualizations/lib/utils";
+import * as Lib from "metabase-lib";
+import Question from "metabase-lib/v1/Question";
+import type Metadata from "metabase-lib/v1/metadata/Metadata";
+import type { ParameterWithTarget } from "metabase-lib/v1/parameters/types";
+import * as ML_Urls from "metabase-lib/v1/urls";
+import type {
+  Card,
+  Dashboard,
+  QuestionDashboardCard,
+} from "metabase-types/api";
+import type { DashboardState } from "metabase-types/store";
+
+/**
+ * All navigation URLs from dashboards to cards (e.x. clicking a title, drill through)
+ * should come from this function, which merges any currently applied dashboard filters
+ * into the new card / URL parameters.
+ */
+export const getNewCardUrl = (
+  metadata: Metadata,
+  dashboardState: DashboardState,
+  {
+    nextCard,
+    previousCard,
+    dashcard,
+    objectId,
+  }: {
+    nextCard: Card;
+    previousCard: Card;
+    dashcard: QuestionDashboardCard;
+    objectId?: number | string;
+  },
+): string | undefined => {
+  const { dashboardId, dashboards, parameterValues } = dashboardState;
+
+  if (dashboardId === null) {
+    return undefined;
+  }
+
+  const dashboard = dashboards[dashboardId];
+  const cardAfterClick = getCardAfterVisualizationClick(nextCard, previousCard);
+
+  let question = new Question(cardAfterClick, metadata);
+  const { isEditable } = Lib.queryDisplayInfo(question.query());
+  if (isEditable) {
+    question = question
+      .setDisplay(cardAfterClick.display || previousCard.display)
+      .setSettings(dashcard.card.visualization_settings)
+      .lockDisplay();
+  } else {
+    question = question.setCard(dashcard.card).setDashboardProps({
+      dashboardId: dashboard.id,
+      dashcardId: dashcard.id,
+    });
+  }
+
+  const parametersMappedToCard = getParametersMappedToDashcard(
+    dashboard.parameters,
+    dashcard,
+  );
+
+  // When drilling from a native model, the drill can return a new question
+  // querying a table for which we don't have any metadata for
+  // When building a question URL, it'll usually clean the query and
+  // strip clauses referencing fields from tables without metadata
+  const previousQuestion = new Question(previousCard, metadata);
+  const { isNative: isPreviousNative } = Lib.queryDisplayInfo(
+    previousQuestion.query(),
+  );
+
+  const isFromModelOrMetric =
+    previousQuestion.type() === "model" || previousQuestion.type() === "metric";
+  const isDrillingFromNativeModelOrMetric =
+    isFromModelOrMetric && isPreviousNative;
+
+  const url = ML_Urls.getUrlWithParameters(
+    question,
+    parametersMappedToCard,
+    parameterValues,
+    {
+      clean: !isDrillingFromNativeModelOrMetric,
+      objectId,
+    },
+  );
+
+  return url;
+};
+
+export function getParametersMappedToDashcard(
+  parameters: Dashboard["parameters"],
```

Comment:
We used to pass an entire `Dashboard` here, but only parameters are needed.
I had do to this change due to `StoreDashboard` not being the same as `Dashboard`.